File: mobile/lib/models/brief.dart
/// Brief Your Agent — client-side models (was Coach — renamed AT:R27).
/// Mirrors backend/app/schemas/brief.py.
library;

enum BriefMode { fromScratch, fromPastCalls, raw }

extension BriefModeJson on BriefMode {
  String get wire {
    switch (this) {
      case BriefMode.fromScratch:
        return 'from_scratch';
      case BriefMode.fromPastCalls:
        return 'from_past_calls';
      case BriefMode.raw:
        return 'raw';
    }
  }
}

class UserOverlay {
  const UserOverlay({
    required this.id,
    required this.userId,
    required this.agentId,
    required this.version,
    required this.content,
    required this.plainEnglish,
    required this.createdAt,
    this.basedOnSession,
  });

  final String id;
  final String userId;
  final String agentId;
  final int version;
  final String content;
  final String plainEnglish;
  final DateTime createdAt;
  final String? basedOnSession;

  factory UserOverlay.fromJson(Map<String, dynamic> j) {
    return UserOverlay(
      id: j['id'] as String,
      userId: j['user_id'] as String,
      agentId: j['agent_id'] as String,
      version: (j['version'] as num).toInt(),
      content: j['content'] as String,
      plainEnglish: j['plain_english'] as String,
      createdAt: DateTime.parse(j['created_at'] as String),
      basedOnSession: j['based_on_session'] as String?,
    );
  }
}

class BriefProposal {
  const BriefProposal({
    required this.id,
    required this.sessionId,
    required this.plainEnglish,
    required this.overlayAddition,
    required this.fullOverlayPreview,
    required this.refused,
    this.refusalReason,
    this.proposedAt,
  });

  final String id;
  final String sessionId;
  final String plainEnglish;
  final String overlayAddition;
  final String fullOverlayPreview;
  final bool refused;
  final String? refusalReason;
  // Backend default-factories this on creation. Nullable on the wire only
  // because old payloads pre-AT:R32 didn't ship it. Once all clients are
  // current, mark non-null. (BL14, AT:R32.)
  final DateTime? proposedAt;

  factory BriefProposal.fromJson(Map<String, dynamic> j) {
    return BriefProposal(
      id: j['id'] as String,
      sessionId: j['session_id'] as String,
      plainEnglish: j['plain_english'] as String? ?? '',
      overlayAddition: j['overlay_addition'] as String? ?? '',
      fullOverlayPreview: j['full_overlay_preview'] as String? ?? '',
      refused: (j['refused'] as bool?) ?? false,
      refusalReason: j['refusal_reason'] as String?,
      proposedAt: j['proposed_at'] != null
          ? DateTime.parse(j['proposed_at'] as String)
          : null,
    );
  }
}

class BriefSession {
  const BriefSession({
    required this.id,
    required this.userId,
    required this.agentId,
    required this.mode,
    required this.locale,
    required this.baseOverlayVersion,
    this.mandateUsed = const {},
    this.pendingProposal,
    this.startedAt,
  });

  final String id;
  final String userId;
  final String agentId;
  final String mode;
  final String locale;
  final int baseOverlayVersion;

  /// Full snapshot of the mandate at session start. Kept on the client for
  /// post-hoc safety-floor audit + offline replay. Opaque map — same shape
  /// as the backend `Mandate.model_dump()`.
  final Map<String, dynamic> mandateUsed;

  /// The proposal currently awaiting accept/reject, if any. Cleared when the
  /// user accepts or rejects.
  final BriefProposal? pendingProposal;

  /// Backend default-factories this when the session is created. Nullable
  /// only for old payloads pre-AT:R32. (BL14, AT:R32.)
  final DateTime? startedAt;

  factory BriefSession.fromJson(Map<String, dynamic> j) {
    return BriefSession(
      id: j['id'] as String,
      userId: j['user_id'] as String,
      agentId: j['agent_id'] as String,
      mode: j['mode'] as String,
      locale: j['locale'] as String? ?? 'en',
      baseOverlayVersion: ((j['base_overlay_version'] as num?) ?? 0).toInt(),
      mandateUsed:
          (j['mandate_used'] as Map?)?.cast<String, dynamic>() ?? const {},
      pendingProposal: j['pending_proposal'] != null
          ? BriefProposal.fromJson(
              (j['pending_proposal'] as Map).cast<String, dynamic>(),
            )
          : null,
      startedAt: j['started_at'] != null
          ? DateTime.parse(j['started_at'] as String)
          : null,
    );
  }
}

class BriefStartResponse {
  const BriefStartResponse({
    required this.session,
    required this.openingMessage,
    this.currentOverlay,
  });

  final BriefSession session;
  final UserOverlay? currentOverlay;
  final String openingMessage;

  factory BriefStartResponse.fromJson(Map<String, dynamic> j) {
    final overlayJson = j['current_overlay'];
    return BriefStartResponse(
      session: BriefSession.fromJson(j['session'] as Map<String, dynamic>),
      currentOverlay: overlayJson == null
          ? null
          : UserOverlay.fromJson(overlayJson as Map<String, dynamic>),
      openingMessage: j['opening_message'] as String? ?? '',
    );
  }
}

class BriefHistory {
  const BriefHistory({
    required this.agentId,
    required this.userId,
    required this.versions,
    required this.activeVersion,
    required this.editCount,
    this.editsRemaining,
  });

  final String agentId;
  final String userId;
  final List<UserOverlay> versions;
  final int activeVersion;
  final int editCount;
  final int? editsRemaining;

  factory BriefHistory.fromJson(Map<String, dynamic> j) {
    final list = (j['versions'] as List<dynamic>?) ?? const [];
    return BriefHistory(
      agentId: j['agent_id'] as String,
      userId: j['user_id'] as String,
      versions: list
          .map((e) => UserOverlay.fromJson(e as Map<String, dynamic>))
          .toList(),
      activeVersion: ((j['active_version'] as num?) ?? 0).toInt(),
      editCount: ((j['edit_count'] as num?) ?? 0).toInt(),
      editsRemaining: j['edits_remaining'] == null
          ? null
          : ((j['edits_remaining'] as num).toInt()),
    );
  }
}

File: backend/app/schemas/brief.py
"""Brief Your Agent schemas.

A Brief session is a conversational editor for the user_overlay block of an
agent's prompt. The user chats with the agent (in briefing mode); when both
sides agree on a change, the agent produces a structured BriefProposal — a
plain-English summary plus a markdown overlay-addition. The user Accepts,
Refines, or Rejects.

Accepted proposals become a new UserOverlay version. Rollback is one tap.

Renamed from "Coach Your Agent" in AT:R27. The previous symbol names
(CoachStartRequest etc.) are kept as deprecated aliases at the bottom of
this module so legacy imports keep working through the deprecation
window.

Spec: docs/02_agents/brief_your_agent.md (renamed from coach_your_agent.md in AT:R31).
Safety floor: docs/02_agents/safety_floor.md (PM mandate enforcement is
un-briefable; see BriefRefusal in brief_engine.py).
"""

from datetime import datetime, timezone
from enum import Enum
from typing import Literal
from uuid import UUID, uuid4

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.agents import AgentId
from app.schemas.one_on_one import ChatMsg


class BriefMode(str, Enum):
    FROM_SCRATCH = "from_scratch"
    FROM_PAST_CALLS = "from_past_calls"
    RAW = "raw"


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class UserOverlay(BaseModel):
    """A versioned overlay block for a single (user, agent) pair.

    `content` is the markdown that gets injected into the prompt between the
    mandate overlay and the safety floor (PM only).

    `plain_english` is the human-facing description shown in the history list.
    """

    model_config = ConfigDict(use_enum_values=True)

    id: UUID = Field(default_factory=uuid4)
    user_id: UUID
    agent_id: AgentId
    version: int = Field(..., ge=0)
    content: str
    plain_english: str
    created_at: datetime = Field(default_factory=_utcnow)
    based_on_session: UUID | None = None


class BriefProposal(BaseModel):
    """A pending proposal from the agent. Generated when user signals agreement.

    The user can Accept / Refine / Reject. Accept turns this into a new
    UserOverlay version.
    """

    model_config = ConfigDict(use_enum_values=True)

    id: UUID = Field(default_factory=uuid4)
    session_id: UUID
    plain_english: str
    overlay_addition: str
    full_overlay_preview: str
    proposed_at: datetime = Field(default_factory=_utcnow)
    refused: bool = False
    refusal_reason: str | None = None


class BriefSession(BaseModel):
    """A live Brief conversation."""

    model_config = ConfigDict(use_enum_values=True)

    id: UUID = Field(default_factory=uuid4)
    user_id: UUID
    agent_id: AgentId
    mode: BriefMode
    mandate_used: dict
    locale: str = "en"
    base_overlay_version: int = 0
    pending_proposal: BriefProposal | None = None
    started_at: datetime = Field(default_factory=_utcnow)


# ── Request / response payloads ────────────────────────────────────────────


class BriefStartRequest(BaseModel):
    model_config = ConfigDict(use_enum_values=True)

    agent_id: AgentId
    user_id: UUID
    mode: BriefMode = BriefMode.FROM_SCRATCH
    mandate_override: dict | None = None
    locale: str = "en"


class BriefStartResponse(BaseModel):
    model_config = ConfigDict(use_enum_values=True)

    session: BriefSession
    current_overlay: UserOverlay | None
    opening_message: str


class BriefMessageRequest(BaseModel):
    session_id: UUID
    user_message: str
    history: list[ChatMsg] = Field(default_factory=list)


class BriefProposeRequest(BaseModel):
    """Ask the agent to crystallise the conversation into a structured proposal."""

    session_id: UUID
    history: list[ChatMsg] = Field(default_factory=list)


class BriefAcceptRequest(BaseModel):
    session_id: UUID
    proposal_id: UUID


class BriefRejectRequest(BaseModel):
    session_id: UUID
    proposal_id: UUID


class BriefRollbackRequest(BaseModel):
    user_id: UUID
    agent_id: AgentId
    to_version: int


class BriefHistoryResponse(BaseModel):
    model_config = ConfigDict(use_enum_values=True)

    agent_id: AgentId
    user_id: UUID
    versions: list[UserOverlay]
    active_version: int
    edit_count: int
    edits_remaining: int | None  # None = unlimited


class BriefRefusal(BaseModel):
    """Returned when Brief refuses a proposal (e.g. PM safety-floor violation,
    mandate compliance, off-role)."""

    reason: Literal[
        "safety_floor",
        "mandate_compliance",
        "off_role",
        "edit_limit_reached",
        "internal_error",
    ]
    message: str
    suggestion: str | None = None


# ── Backward-compat aliases (AT:R27 rename) ──────────────────────────────
# Legacy code importing `CoachX` continues to work through the deprecation
# window. Remove these once Flutter +17 testers are retired and all internal
# call sites have been swept.

CoachMode = BriefMode
CoachProposal = BriefProposal
CoachSession = BriefSession
CoachStartRequest = BriefStartRequest
CoachStartResponse = BriefStartResponse
CoachMessageRequest = BriefMessageRequest
CoachProposeRequest = BriefProposeRequest
CoachAcceptRequest = BriefAcceptRequest
CoachRejectRequest = BriefRejectRequest
CoachRollbackRequest = BriefRollbackRequest
CoachHistoryResponse = BriefHistoryResponse
CoachRefusal = BriefRefusal
